File: Assignment 2/task1.scala
import org.apache.spark.{SparkConf, SparkContext}
import scala.collection.mutable.{ArrayBuffer, HashMap}
import scala.util.Try
import java.io.PrintWriter

object task1 {

  // Hash function similar to Python's xor_hash:
  def xorHash(a: String, b: String, numBuckets: Int): Int = {
    val aa = Try(a.toInt).getOrElse(a.hashCode)
    val bb = Try(b.toInt).getOrElse(b.hashCode)
    ((aa ^ bb).abs) % numBuckets
  }

  // Build baskets for Case 1: user -> set(businesses)
  def buildBasketsCase1(rdd: org.apache.spark.rdd.RDD[Array[String]]): org.apache.spark.rdd.RDD[Set[String]] = {
    rdd.map(row => (row(0), row(1)))
       .groupByKey()
       .map { case (_, items) => items.toSet }
  }

  // Build baskets for Case 2: business -> set(users)
  def buildBasketsCase2(rdd: org.apache.spark.rdd.RDD[Array[String]]): org.apache.spark.rdd.RDD[Set[String]] = {
    rdd.map(row => (row(1), row(0)))
       .groupByKey()
       .map { case (_, users) => users.toSet }
  }

  // Optimized A-Priori candidate generation using prefix grouping.
  def aprioriGen(prevFreq: List[Seq[String]], k: Int): List[Seq[String]] = {
    // Group (k-1)-itemsets by their first k-2 items.
    val grouped = prevFreq.groupBy(itemset => itemset.take(k - 2))
    val candidates = ArrayBuffer[Seq[String]]()
    val prevFreqSet = prevFreq.toSet
    for ((prefix, group) <- grouped) {
      if (group.size >= 2) {
        // Sort the group lexicographically using mkString.
        val sortedGroup = group.sortBy(x => x.mkString(","))
        for (i <- sortedGroup.indices; j <- i + 1 until sortedGroup.size) {
          val candidate = (sortedGroup(i) ++ Seq(sortedGroup(j).last)).sorted
          // Prune candidate: every (k-1)-subset must be frequent.
          if (candidate.combinations(k - 1).forall(subset => prevFreqSet.contains(subset))) {
            candidates += candidate
          }
        }
      }
    }
    candidates.toList
  }

  // Local PCY pass (SON Phase 1) on a partition.
  // Returns an iterator of (candidate itemset, 1)
  def localPcy(partBaskets: Iterator[Set[String]], totalCount: Long, globalSupport: Double, numBuckets: Int = 1000): Iterator[(Seq[String], Int)] = {
    val baskets = partBaskets.toList
    val partSize = baskets.size
    if (partSize == 0) return Iterator.empty

    // Local support threshold proportional to the partition.
    val fraction = partSize.toDouble / totalCount.toDouble
    val localThreshFloat = globalSupport * fraction
    val localThreshold = if (localThreshFloat < 1) 1 else math.ceil(localThreshFloat).toInt

    // Count singletons and hash pairs.
    val itemCount = new HashMap[String, Int]()
    val hashCount = new HashMap[Int, Int]()
    baskets.foreach { basket =>
      val basketList = basket.toList
      basketList.foreach { item =>
        itemCount(item) = itemCount.getOrElse(item, 0) + 1
      }
      for (i <- 0 until basketList.length - 1) {
        for (j <- i + 1 until basketList.length) {
          val idx = xorHash(basketList(i), basketList(j), numBuckets)
          hashCount(idx) = hashCount.getOrElse(idx, 0) + 1
        }
      }
    }

    // Build a bitmap from hash counts.
    val bitMap = Array.fill(numBuckets)(0)
    hashCount.foreach { case (idx, count) =>
      if (count >= localThreshold) bitMap(idx) = 1
    }

    // Frequent singletons.
    val freqSingles = itemCount.filter { case (_, cnt) => cnt >= localThreshold }.keys.toList.sorted
    val localCandidates = ArrayBuffer[(Seq[String], Int)]()
    freqSingles.foreach(item => localCandidates += ((Seq(item), 1)))

    // Candidate pairs from frequent singletons.
    val pairCandidates = ArrayBuffer[(String, String)]()
    for (i <- freqSingles.indices; j <- i + 1 until freqSingles.length)
      pairCandidates += ((freqSingles(i), freqSingles(j)))
    val goodPairs = pairCandidates.filter { case (a, b) =>
      val idx = xorHash(a, b, numBuckets)
      bitMap(idx) == 1
    }
    if (goodPairs.isEmpty) return localCandidates.iterator

    // Remove non-frequent singles from baskets.
    val freqSinglesSet = freqSingles.toSet
    val updatedBaskets = baskets.map(b => b.filter(freqSinglesSet.contains).toList)

    // Exact counting for candidate pairs.
    val pairCount = new HashMap[(String, String), Int]()
    goodPairs.foreach(pair => pairCount(pair) = 0)
    updatedBaskets.foreach { bList =>
      val bSet = bList.toSet
      pairCount.keys.foreach { case pair =>
        if (bSet.contains(pair._1) && bSet.contains(pair._2))
          pairCount(pair) = pairCount(pair) + 1
      }
    }
    val freqPairs = goodPairs.filter(pair => pairCount(pair) >= localThreshold)
    freqPairs.foreach { pair =>
      localCandidates += ((Seq(pair._1, pair._2).sorted, 1))
    }
    if (freqPairs.isEmpty) return localCandidates.iterator

    // For k >= 3, use the optimized aprioriGen.
    var currentSets: List[Seq[String]] = freqPairs.map { case (a, b) => Seq(a, b) }.toList
    var k = 3
    while (currentSets.nonEmpty) {
      val candidateK = aprioriGen(currentSets, k)
      if (candidateK.isEmpty) {
        currentSets = Nil
      } else {
        val candCount = new HashMap[Seq[String], Int]()
        candidateK.foreach(c => candCount(c) = 0)
        updatedBaskets.foreach { bList =>
          val bSet = bList.toSet
          candidateK.foreach { candidate =>
            if (candidate.toSet.subsetOf(bSet))
              candCount(candidate) = candCount(candidate) + 1
          }
        }
        val freqNew = candCount.filter { case (_, cnt) => cnt >= localThreshold }.keys.toList
        freqNew.foreach(c => localCandidates += ((c, 1)))
        currentSets = freqNew
        k += 1
      }
    }

    localCandidates.iterator
  }

  // SON Phase 2: Global candidate counting.
  def pass2Count(basketsIter: Iterator[Set[String]], candList: Seq[Seq[String]]): Iterator[(Seq[String], Int)] = {
    val outMap = new HashMap[Seq[String], Int]()
    basketsIter.foreach { basket =>
      val bSet = basket  // already a set
      candList.foreach { candidate =>
        if (candidate.toSet.subsetOf(bSet))
          outMap(candidate) = outMap.getOrElse(candidate, 0) + 1
      }
    }
    outMap.iterator
  }

  // Group and sort itemsets by size.
  def groupAndSortItemsets(itemsets: Seq[Seq[String]]): Seq[(Int, Seq[Seq[String]])] = {
    val groupMap = itemsets.groupBy(_.length)
    groupMap.toSeq.sortBy(_._1).map { case (size, items) =>
      (size, items.sortBy(itemset => itemset.mkString(",")))
    }
  }

  // Format an itemset as ("'a'", "'b'", ...)
  def formatItemset(itup: Seq[String]): String = {
    "(" + itup.map(item => s"'$item'").mkString(", ") + ")"
  }

  // Write out results.
  def writeResults(candGrouped: Seq[(Int, Seq[Seq[String]])],
                   freqGrouped: Seq[(Int, Seq[Seq[String]])],
                   outFile: String): Unit = {
    val writer = new PrintWriter(outFile)
    try {
      writer.println("Candidates:")
      candGrouped.foreach { case (_, candidates) =>
        writer.println(candidates.map(formatItemset).mkString(","))
        writer.println()
      }
      writer.println("Frequent Itemsets:")
      freqGrouped.foreach { case (_, freqItems) =>
        writer.println(freqItems.map(formatItemset).mkString(","))
        writer.println()
      }
    } finally {
      writer.close()
    }
  }

  def main(args: Array[String]): Unit = {
    if (args.length != 4) {
      println("Usage: spark-submit --class task1 <case_number> <support> <input_file> <output_file>")
      sys.exit(1)
    }

    val caseNum = args(0).toInt
    val support = args(1).toDouble
    val inputFile = args(2)
    val outputFile = args(3)

    val conf = new SparkConf().setAppName("Task1SON")
      .set("spark.executor.memory", "4g")
      .set("spark.driver.memory", "4g")
    val sc = new SparkContext(conf)
    sc.setLogLevel("ERROR")

    val startT = System.nanoTime()

    // Build baskets and cache them.
    val lines = sc.textFile(inputFile)
    val header = lines.first()
    val dataRdd = lines.filter(line => line != header).map(line => line.trim.split(","))
    val basketsRdd = if (caseNum == 1) buildBasketsCase1(dataRdd) else buildBasketsCase2(dataRdd)
    basketsRdd.cache()
    val totalCount = basketsRdd.count()

    // SON Phase 1: Local candidate generation.
    val localCandidates = basketsRdd
      .mapPartitions(part => localPcy(part, totalCount, support, 1000))
      .reduceByKey(_ + _)
      .map(_._1)
      .distinct()
      .collect()
      .toSeq

    val candGrouped = groupAndSortItemsets(localCandidates)

    // Broadcast candidate list for SON Phase 2.
    val bcCandList = sc.broadcast(localCandidates)

    // SON Phase 2: Global count.
    val freqCounts = basketsRdd
      .mapPartitions(part => pass2Count(part, bcCandList.value))
      .reduceByKey(_ + _)
      .filter { case (_, count) => count >= support }
      .map(_._1)
      .collect()
      .toSeq

    val freqGrouped = groupAndSortItemsets(freqCounts)

    writeResults(candGrouped, freqGrouped, outputFile)

    val endT = System.nanoTime()
    println(s"Duration: ${(endT - startT) / 1e9}")
    sc.stop()
  }
}

File: Assignment 2/task1.py
#!/usr/bin/env python3
import sys
import time
from itertools import combinations
from collections import defaultdict
from pyspark import SparkContext

###############################################################################
#                              HASH FUNCTION                                  #
###############################################################################

def xor_hash(a, b, num_buckets):
    """
    Convert 'a','b' to int if possible, else fallback to hash,
    then (x ^ y) % num_buckets.
    """
    try:
        aa = int(a)
        bb = int(b)
    except ValueError:
        aa = hash(a)
        bb = hash(b)
    return (aa ^ bb) % num_buckets

###############################################################################
#                      BUILD BASKETS FOR CASE1 / CASE2                        #
###############################################################################

def build_baskets_case1(rdd):
    """
    Case 1: (user -> set(businesses)).
    Return RDD of sets of business IDs.
    """
    return (rdd
            .map(lambda row: (row[0], row[1]))
            .groupByKey()
            .mapValues(set)
            .map(lambda x: x[1]))

def build_baskets_case2(rdd):
    """
    Case 2: (business -> set(users)).
    Return RDD of sets of user IDs.
    """
    return (rdd
            .map(lambda row: (row[1], row[0]))
            .groupByKey()
            .mapValues(set)
            .map(lambda x: x[1]))

###############################################################################
#                             LOCAL PCY PASS                                  #
###############################################################################

def local_pcy(part_baskets, total_count, global_support, num_buckets=1000):
    """
    SON local pass (PCY) with fraction-based local threshold:
      local_threshold = max(1, support * (len_partition/total_count)).

    Steps:
      - Count singletons; build pair-hash counts => build a bitmap
      - Identify freq singles, candidate pairs
      - Remove non-freq singles from baskets & count pairs exactly
      - Extend to k>=3 via naive combo approach.

    Returns [(itemset, 1), ...] representing local frequent sets.
    """
    baskets = list(part_baskets)
    part_size = len(baskets)
    if part_size == 0:
        return []

    # local threshold
    fraction = part_size / float(total_count)
    local_thresh_float = global_support * fraction
    if local_thresh_float < 1:
        local_thresh_float = 1.0
    local_threshold = local_thresh_float

    # Pass1: count singletons, hash pairs
    item_count = {}
    hash_count = {}

    for basket_set in baskets:
        basket_list = list(basket_set)

        # count each item
        for itm in basket_list:
            item_count[itm] = item_count.get(itm, 0) + 1

        # hash each pair
        for i in range(len(basket_list) - 1):
            for j in range(i + 1, len(basket_list)):
                idx = xor_hash(basket_list[i], basket_list[j], num_buckets)
                hash_count[idx] = hash_count.get(idx, 0) + 1

    # build a bit_map from hash_count
    bit_map = [0]*num_buckets
    for idx, val in hash_count.items():
        if val >= local_threshold:
            bit_map[idx] = 1

    # gather freq singletons
    freq_singles = []
    for it, cnt in item_count.items():
        if cnt >= local_threshold:
            freq_singles.append(it)

    local_candidates = []
    # add singletons as (tuple([s]), 1)
    for s in freq_singles:
        local_candidates.append((tuple([s]), 1))

    # build candidate pairs from freq singles
    pair_candidates = []
    for i in range(len(freq_singles)):
        for j in range(i+1, len(freq_singles)):
            pair_candidates.append((freq_singles[i], freq_singles[j]))

    # remove pairs whose hashed bucket isn't frequent
    good_pairs = []
    for pr in pair_candidates:
        h_idx = xor_hash(pr[0], pr[1], num_buckets)
        if bit_map[h_idx] == 1:
            good_pairs.append(pr)

    if not good_pairs:
        # no freq pairs => done
        return local_candidates

    # remove from baskets the non-freq singles (for exact counting of pairs+)
    freq_singles_set = set(freq_singles)
    updated_baskets = []
    for bset in baskets:
        # keep only freq singles
        keep_items = [x for x in bset if x in freq_singles_set]
        updated_baskets.append(keep_items)

    # exact count pairs
    pair_count = dict.fromkeys(good_pairs, 0)
    for b in updated_baskets:
        b_as_set = set(b)
        for p_tup in pair_count.keys():
            if p_tup[0] in b_as_set and p_tup[1] in b_as_set:
                pair_count[p_tup] += 1

    freq_pairs = []
    for pr, cval in pair_count.items():
        if cval >= local_threshold:
            freq_pairs.append(pr)
            local_candidates.append((tuple(sorted(pr)), 1))

    if not freq_pairs:
        return local_candidates

    # k>=3 combos
    current_sets = freq_pairs
    k = 3
    while True:
        # gather distinct items from freq (k-1)-sets
        all_items = set()
        for cset in current_sets:
            for x in cset:
                all_items.add(x)

        # build combos of size k
        cands_k = list(combinations(sorted(all_items), k))
        cand_count = dict.fromkeys(cands_k, 0)

        for b_list in updated_baskets:
            bset = set(b_list)
            for ck in cands_k:
                if set(ck).issubset(bset):
                    cand_count[ck] += 1

        freq_new = []
        for ccombo, ccount in cand_count.items():
            if ccount >= local_threshold:
                freq_new.append(ccombo)
                local_candidates.append((tuple(ccombo), 1))

        if not freq_new:
            break
        current_sets = freq_new
        k += 1

    return local_candidates

###############################################################################
#                              SON PASS 2                                     #
###############################################################################

def pass2_count(basket_iter, cand_list):
    """
    For each basket, check each candidate. If candidate is subset of basket, emit (cand,1).
    """
    out_map = defaultdict(int)
    for bset in basket_iter:
        b_as_set = set(bset)
        for ctuple in cand_list:
            if set(ctuple).issubset(b_as_set):
                out_map[ctuple] += 1
    return out_map.items()

###############################################################################
#                          OUTPUT FORMATTING                                  #
###############################################################################

def group_and_sort_itemsets(itemsets):
    """
    Group itemsets by size -> sort each group in lexicographical order 
    (treat each element as string).
    
    Returns a list of (size, [list_of_tuples_sorted]) in ascending size order.
    """
    from collections import defaultdict
    group_map = defaultdict(list)
    for it in itemsets:
        group_map[len(it)].append(it)

    # Build a sorted list of (size, [itemsets]).
    results = []
    for sz in sorted(group_map.keys()):
        # Sort itemsets of size=sz lexicographically as strings
        group_map[sz].sort(key=lambda tup: tuple(str(x) for x in tup))
        results.append((sz, group_map[sz]))
    return results

def format_itemset(itup):
    """
    Convert a tuple like ('2', '6') into "('2', '6')" 
    or a single tuple ('2',) into "('2')".
    Each item is wrapped in single quotes and a space is added after each comma.
    """
    return "(" + ", ".join("'" + str(x) + "'" for x in itup) + ")"

def write_results(cand_grouped, freq_grouped, out_file):
    """
    Print all candidates and frequent itemsets in the required grouped format:

    Candidates:
    ('i1', 'i2'),('i3', 'i4')...

    ('i5')...

    Frequent Itemsets:
    ('i1', 'i2'),('i3', 'i4')...

    etc.

    Each size's itemsets are on ONE line, separated by commas, then a blank line.
    """
    with open(out_file, "w") as f:
        # ---------------------
        # Candidates
        # ---------------------
        f.write("Candidates:\n")
        for (sz, cands) in cand_grouped:
            # put all cands of this size on ONE line, separated by commas
            line_str = ",".join(format_itemset(c) for c in cands)
            f.write(line_str + "\n\n")  # blank line after

        # ---------------------
        # Frequent Itemsets
        # ---------------------
        f.write("Frequent Itemsets:\n")
        for (sz, frs) in freq_grouped:
            line_str = ",".join(format_itemset(fset) for fset in frs)
            f.write(line_str + "\n\n")  # blank line after each size group

###############################################################################
#                                   MAIN                                      #
###############################################################################

def main():
    if len(sys.argv) != 5:
        print("Usage: spark-submit task1.py <case_number> <support> <input_file> <output_file>")
        sys.exit(-1)

    case_num = int(sys.argv[1])
    support = float(sys.argv[2])
    input_file = sys.argv[3]
    output_file = sys.argv[4]

    sc = SparkContext(appName="task1_updated_code")
    sc.setLogLevel("ERROR")

    start_t = time.time()

    lines = sc.textFile(input_file)
    header = lines.first()  # handle a potential header row
    data_rdd = lines.filter(lambda x: x != header).map(lambda x: x.strip().split(","))

    # Build baskets depending on case #
    if case_num == 1:
        baskets_rdd = build_baskets_case1(data_rdd)
    else:
        baskets_rdd = build_baskets_case2(data_rdd)

    total_count = baskets_rdd.count()

    # ------------------
    # SON Phase 1
    # ------------------
    local_candidates = (
        baskets_rdd
        .mapPartitions(lambda part: local_pcy(part, total_count, support, 1000))
        .reduceByKey(lambda a, b: a + b)
        .map(lambda x: x[0])
        .distinct()
        .collect()
    )

    # group candidates by size and sort them
    cand_grouped = group_and_sort_itemsets(local_candidates)

    # ------------------
    # SON Phase 2
    # ------------------
    cand_list = list(local_candidates)  # broadcast to pass2
    freq_counts = (
        baskets_rdd
        .mapPartitions(lambda p: pass2_count(p, cand_list))
        .reduceByKey(lambda x, y: x + y)
        .filter(lambda x: x[1] >= support)
        .map(lambda x: x[0])
        .collect()
    )

    freq_grouped = group_and_sort_itemsets(freq_counts)

    # ------------------
    # Write out results
    # ------------------
    write_results(cand_grouped, freq_grouped, output_file)

    end_t = time.time()
    print("Duration:", end_t - start_t)

    sc.stop()

if __name__ == "__main__":
    main()
